### PR TITLE
Run step for soca (grid generation and 3dvar fgat)

### LINKED ISSUE #29: Test the run step as part of the ci testing

### PR DESCRIPTION
**Pull request by guillaumevernieres**
## Description

This PR adds the necessary features to allow testing of:
- The `soca` grid generation 
- The generation of the flow dependent parametric diagonal of B
- 3DVARFGAT with a correlation operator set to Id

All of the above are executed within [exgdas_global_marine_analysis_run.sh](https://github.com/NOAA-EMC/GDASApp/blob/feature/soca-run/scripts/exgdas_global_marine_analysis_run.sh). 
The general idea is to attempt to replicate the environment that will be inherited from the global workflow and run the `prep` and `run` scripts with toy backgrounds within `ctests`. I don't expect the `prep` or `run` steps to work as is within the global workflow. That specific work will be the subject of an other PR.

## Features added:
- The [exgdas_global_marine_analysis_prep.py](https://github.com/NOAA-EMC/GDASApp/blob/feature/soca-run/scripts/exgdas_global_marine_analysis_prep.py) scripts does the following:
  - Creates the directories to run stuff in and dump outputs into
  - prepares all the static files
  - fetch the obs
  - Link the [gridgen.yaml](https://github.com/NOAA-EMC/GDASApp/blob/feature/soca-run/parm/soca/gridgen/gridgen.yaml) used to run the grid generation application
  - Generates the 3DVAR FGAT `yaml` files, linking to the 3 hourly MOM6 backgrounds
- The [exgdas_global_marine_analysis_run.sh](https://github.com/NOAA-EMC/GDASApp/blob/feature/soca-run/scripts/exgdas_global_marine_analysis_run.sh) runs `soca_gridgen.x` and `soca_var.x`
- A ctest for `exgdas_global_marine_analysis_run.sh`:
  - It is supposedly mimicking what would happen when ran with the global workflow. But it's currently probably quite naive, and maybe some of it is wrong ... The mostly guessed environment is in [runtime_vars.sh](https://github.com/NOAA-EMC/GDASApp/blob/feature/soca-run/test/soca/runtime_vars.sh)
  - It is staging all the restarts and required static files from the `soca` repo, which uses a very low-res test case (5deg)

## Testing
The new features have been tested on `Hera`, 'Orion`, my personal laptop and desktop.

 ## Issue addressed
- fixes #29

### CODE REVIEW

**Comment by guillaumevernieres** [comment]
@CoryMartin-NOAA , @RussTreadon-NOAA , it is still a draft PR, but I'm marking as ready to discuss. 
I was not able to make it work on `Orion`:
- I'm having weird issues with python
- The `run_step` `ctests` is failing in the `ci`. Although the grid generation runs to completion but it is failing when checking for the return value (?):
```console
  terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_M_construct null not valid
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_M_construct null not valid
srun: error: Orion-02-62: tasks 0-1: Aborted
srun: launch/slurm: _step_signal: Terminating StepId=5723194.0
```
but it does terminate with a `status = 0`:
```console
# See here on Orion:
/work2/noaa/stmp/cmartin/CI/GDASApp/PR/62/GDASApp/build/test/soca/3dvar/analysis/exgdas_global_marine_analysis_run.log

[...]
OOPS_STATS ---------------------------------- Parallel Timing Statistics (   2 MPI tasks) -----------------------------------

OOPS_STATS Run end                                  - Runtime:      4.88 sec,  Memory: total:   529.69 Mb, per task: min =   259.68 Mb, max =   270.01 Mb
Run: Finishing soca::GridGen< with status = 0
```

I'll look it this issue more closely tomorrow.

**Review by FanHan-NOAA — DISMISSED** [review]
Checkout the branch and all ctests passed on Orion. Code changes look good to me, although I'm not an expert on soca. Just out of curiosity, is it possible to run a standalone marine analysis with `run_jedi_exe.py`?

**Comment by guillaumevernieres** [comment]
> Checkout the branch and all ctests passed on Orion. Code changes look good to me, although I'm not an expert on soca. Just out of curiosity, is it possible to run a standalone marine analysis with `run_jedi_exe.py`?

Thanks for your review @FanHan-NOAA . 
It isn't possible (yet?). But, the ctests that I added do run 1 cycle of a marine analysis at very low res (5deg). One ctest runs the [marine prep script](https://github.com/NOAA-EMC/GDASApp/blob/feature/soca-run/scripts/exgdas_global_marine_analysis_prep.py) and the subsequent one runs the [marine run script](https://github.com/NOAA-EMC/GDASApp/blob/feature/soca-run/scripts/exgdas_global_marine_analysis_run.sh) (where the JEDI DA is called).
This is probably how `run_jedi_exe.py` should be refactored, calls to the prep and run steps, and probably added to our ctests.

File: parm/soca/obs/config/adt_j3.yaml [hunk 1]
```diff
@@ -1,9 +1,9 @@
 obs space:
   name: adt_j3
   obsdatain:
-    obsfile: ./obs/adt_j3.{{window_begin}}.nc4
+    obsfile: ./obs/adt_j3.nc4
```
Comment (CoryMartin-NOAA) [inline parm/soca/obs/config/adt_j3.yaml]: We should conform to the same file naming convention here, right?

File: parm/soca/obs/config/adt_j3.yaml [hunk 2]
```diff
@@ -1,9 +1,9 @@
...
   obsdataout:
-    obsfile: ./obs_out/adt_j3.{{window_begin}}.nc4
+    obsfile: ./obs_out/adt_j3.$(OBS_DATE).nc4
```
Comment (CoryMartin-NOAA) [inline parm/soca/obs/config/adt_j3.yaml]: same comment, this should probably be consistent across components?

File: parm/soca/obs/config/sst_noaa19_l3u.yaml [hunk 3]
```diff
@@ -1,9 +1,9 @@
 obs space:
   name: sst_noaa19_l3u
   obsdatain:
-    obsfile: ./obs/sst_noaa19_l3u.{{window_begin}}.nc4
+    obsfile: ./obs/sst_noaa19_l3u.nc4
   obsdataout:
-    obsfile: ./obs_out/sst_noaa19_l3u.{{window_begin}}.nc4
+    obsfile: ./obs_out/sst_noaa19_l3u.$(OBS_DATE).nc4
```
Comment (CoryMartin-NOAA) [inline parm/soca/obs/config/sst_noaa19_l3u.yaml]: Same comment for these two lines as above.

File: scripts/exgdas_global_marine_analysis_prep.py [hunk 4]
```diff
@@ -59,16 +59,23 @@ def gen_bkg_list(bkg_path='.', file_type='MOM', yaml_name='bkg.yaml'):
...
 comout = os.getenv('COMOUT')
 comin_obs = os.getenv('COMIN_OBS')
+staticsoca_dir = os.getenv('SOCA_INPUT_FIX_DIR')
 
 # create analysis directory for files
 anl_dir = os.path.join(comout, 'analysis')
 ufsda.mkdir(anl_dir)
 
+# create output directory for obs
+ufsda.mkdir(os.path.join(comout, 'analysis', 'obs_out'))
```
Comment (CoryMartin-NOAA) [inline scripts/exgdas_global_marine_analysis_prep.py]: "diags"

**Review by CoryMartin-NOAA — COMMENTED** [review]
Couple of comments but mostly looks good to me.

**Comment by guillaumevernieres** [comment]
@CoryMartin-NOAA , I addressed your comments but the CI is failing. I'll look into it today again, but the last commit I push was testing fine on `Orion`.

**Comment by CoryMartin-NOAA** [comment]
@guillaumevernieres ok, it could also be a CI issue. I can take a look later today as well.

**Comment by guillaumevernieres** [comment]
> @guillaumevernieres ok, it could also be a CI issue. I can take a look later today as well.

@CoryMartin-NOAA ... `soca` develop is failing now. We should probably do something similar to what @travissluka did for `soca` and `soca-science` and have the latest working develop saved in a daily release branch ... one day ...

**Comment by travissluka** [comment]
FYI @guillaumevernieres  https://github.com/JCSDA-internal/soca/issues/758, I should be able to fix it when lorenz goes down for a nap today

**Review by FanHan-NOAA — APPROVED** [review]
Cloned a new GDASApp, built with fv3-bundle, and all ctests passed for this branch on Orion. Changes look good to me, thanks!

**Review by CoryMartin-NOAA — APPROVED** [review]

**Pull request merged**